### PR TITLE
Introducing shared_ptr instead self-implemented of refcounted objects.

### PR DESCRIPTION
**Pull request by pirogronian**
Second commit increased CPU load by about 10% on my machine. Should I proceed or should there  be a better implementation?

### CODE REVIEW

File: src/celengine/timeline.cpp [hunk 1]
```diff
@@ -27,9 +27,7 @@ Timeline::~Timeline()
     for (auto phase : phases)
     {
         // Remove the phase from whatever phase tree contains it.
-        phase->getFrameTree()->removeChild(phase);
-
-        phase->release();
+        phase->getFrameTree()->removeChild(phase.get());
```
Comment (375gnu) [inline src/celengine/timeline.cpp]: I believe  FrameTree owns its children so it should accept sharedptr not raw.

File: src/celengine/timeline.cpp [hunk 2]
```diff
@@ -46,14 +44,13 @@ Timeline::appendPhase(TimelinePhase* phase)
             return false;
     }
 
-    phase->addRef();
-    phases.push_back(phase);
+    phases.push_back(shared_ptr<TimelinePhase>(phase));
```
Comment (375gnu) [inline src/celengine/timeline.cpp]: have you tried to leave `phases.push_back(phase);`? compiler should generate calls to create shared_ptr

File: src/celengine/timeline.h [hunk 3]
```diff
@@ -38,7 +39,7 @@ class Timeline
     void markChanged();
 
 private:
-    std::vector<TimelinePhase*> phases;
+    std::vector<std::shared_ptr<TimelinePhase> > phases;
```
Comment (375gnu) [inline src/celengine/timeline.h]: in modern c++ it's not required to add a space between > and > in templates

File: src/celengine/timelinephase.cpp [hunk 4]
```diff
@@ -34,19 +34,14 @@ TimelinePhase::TimelinePhase(Body* _body,
...
-    refCount(0)
+    m_owner(_owner)
 {
     // assert(owner == orbitFrame->getCenter()->getFrameTree());
-    m_orbitFrame->addRef();
-    m_bodyFrame->addRef();
 }
 
 
 TimelinePhase::~TimelinePhase()
```
Comment (375gnu) [inline src/celengine/timelinephase.cpp]: `~TimelinePhase() = default` in header file

File: src/celengine/timeline.cpp (hunk 1 above)
Comment (pirogronian) [reply src/celengine/timeline.cpp]: I believe it's done this way in second commit.

File: src/celengine/timeline.cpp (hunk 2 above)
Comment (pirogronian) [reply src/celengine/timeline.cpp]: I'm nearly sure 've tried, also hoping compiler will do the rest.

File: src/celengine/timeline.h (hunk 3 above)
Comment (pirogronian) [reply src/celengine/timeline.h]: Good to know. :-)

**Comment by 375gnu** [comment]
> Second commit increased CPU load by about 10% on my machine. Should I proceed or should there be a better implementation?

Damn, i didn't expected so high growth.

Can we live with the current code?

**Comment by pirogronian** [comment]
> Damn, i didn't expected so high growth.

I believe it's due to pass shared_ptr as value. I try to make it passing as a reference.

> Can we live with the current code?

Yes, but also mainteining current set of features. So, good bye dynamic unloading...

**Comment by 375gnu** [comment]
> Yes, but also mainteining current set of features. So, good bye dynamic unloading...

Why you can't unload objects with current ref counters?

**Comment by pirogronian** [comment]
> Why you can't unload objects with current ref counters?

I can and I tried it in this PR. I only wanted to make common base for such a ref counters. You know the rest :-)

**Comment by pirogronian** [comment]
With passing shared_ptr as reference, even with additional shared_ptr usage in Timeline, CPU load seems to remain normal.

File: src/celengine/frametree.cpp [hunk 5]
```diff
@@ -172,23 +170,22 @@ FrameTree::addChild(TimelinePhase* phase)
...
-        (*iter)->release();
         children.erase(iter);
         markChanged();
     }
 }
 
 
 /*! Return the child at the specified index. */
-TimelinePhase*
+const TimelinePhase*
```
Comment (375gnu) [inline src/celengine/frametree.cpp]: why do you have raw pointer here?

i suppose we *should* always return smart pointers, we *can* accept raw pointers (if we don't store them anywhere)

File: src/celengine/frametree.h [hunk 6]
```diff
@@ -39,9 +40,9 @@ class FrameTree
 
     ReferenceFrame* getDefaultReferenceFrame() const;
 
-    void addChild(TimelinePhase* phase);
-    void removeChild(TimelinePhase* phase);
-    TimelinePhase* getChild(unsigned int n) const;
+    void addChild(std::shared_ptr<const TimelinePhase> &phase);
+    void removeChild(std::shared_ptr<const TimelinePhase> &phase);
+    const TimelinePhase* getChild(unsigned int n) const;
```
Comment (375gnu) [inline src/celengine/frametree.h]: and raw pointer here

File: src/celengine/solarsys.cpp [hunk 7]
```diff
@@ -518,7 +509,7 @@ static bool CreateTimeline(Body* body,
         const Timeline* timeline = body->getTimeline();
         if (timeline->phaseCount() == 1)
         {
-            const TimelinePhase* phase = timeline->getPhase(0);
+            auto phase = timeline->getPhase(0);
```
Comment (375gnu) [inline src/celengine/solarsys.cpp]: here you can use raw pointers to speedup a little bit:
`auto phase = timeline->getPhase(0).get();`

File: src/celestia/celx_phase.cpp [hunk 8]
```diff
@@ -28,12 +28,10 @@ class PhaseReference
     PhaseReference(const TimelinePhase& _phase) :
     phase(&_phase)
 {
-        phase->addRef();
```
Comment (375gnu) [inline src/celestia/celx_phase.cpp]: Actually you should remove PhaseReference and use shared_ptr instead directly

File: src/celengine/frametree.cpp (hunk 5 above)
Comment (pirogronian) [reply src/celengine/frametree.cpp]: Probably because I wanted to check impact on perfomance on the first place. Now I'm implementing it more comprehently and I'm stuck with domino effect... Probably will be finished tomorrow.

**Comment by pirogronian** [comment]
After 246aa4c CPU still remains normal. However, lua interface for frames is not tested yet.

**Comment by pirogronian** [comment]
CPU usage still normal :-)

File: src/celengine/frame.cpp [hunk 9]
```diff
@@ -728,8 +700,6 @@ FrameVector::FrameVector(FrameVectorType t) :
 
 FrameVector::~FrameVector()
```
Comment (375gnu) [inline src/celengine/frame.cpp]: can be replaced with `= default`in .h file

File: src/celengine/observer.cpp [hunk 10]
```diff
@@ -816,16 +814,15 @@ void Observer::setFrame(ObserverFrame::CoordinateSystem cs, const Selection& ref
...
  *  universal coordinates will not change.
  */
-void Observer::setFrame(const ObserverFrame& f)
+void Observer::setFrame(const shared_ptr<const ObserverFrame>& f)
 {
-    if (frame != &f)
+    if (frame != f)
     {
-        auto* newFrame = new ObserverFrame(f);
+        auto newFrame = make_shared<ObserverFrame>(*f);
```
Comment (375gnu) [inline src/celengine/observer.cpp]: `auto* newFrame = new ObserverFrame(f);` was required due to ownership and `new`/`delete` issues, now with `shared_ptr` we don't need this and instead of `newFrame` `f` should be used.

File: src/celengine/parseobject.h [hunk 11]
```diff
@@ -46,12 +47,12 @@ RotationModel* CreateRotationModel(Hash* rotationData,
 
 RotationModel* CreateDefaultRotationModel(double syncRotationPeriod);
 
-ReferenceFrame* CreateReferenceFrame(const Universe& universe,
+std::shared_ptr<const ReferenceFrame> CreateReferenceFrame(const Universe& universe,
```
Comment (375gnu) [inline src/celengine/parseobject.h]: why `const` here?

File: src/celengine/parseobject.h [hunk 12]
```diff
@@ -46,12 +47,12 @@ RotationModel* CreateRotationModel(Hash* rotationData,
...
                                      Value* frameValue,
                                      const Selection& defaultCenter,
                                      Body* defaultObserver);
 
-TwoVectorFrame* CreateTopocentricFrame(const Selection& center,
+std::shared_ptr<const TwoVectorFrame> CreateTopocentricFrame(const Selection& center,
```
Comment (375gnu) [inline src/celengine/parseobject.h]: and here?

File: src/celengine/timelinephase.h [hunk 13]
```diff
@@ -42,7 +41,7 @@ class TimelinePhase
         return m_endTime;
     }
 
-    ReferenceFrame* orbitFrame() const
+    const std::shared_ptr<const ReferenceFrame>& orbitFrame() const
```
Comment (375gnu) [inline src/celengine/timelinephase.h]: why `const` here?

File: src/celengine/timelinephase.h [hunk 14]
```diff
@@ -52,7 +51,7 @@ class TimelinePhase
         return m_orbit;
     }
 
-    ReferenceFrame* bodyFrame() const
+    const std::shared_ptr<const ReferenceFrame>& bodyFrame() const
```
Comment (375gnu) [inline src/celengine/timelinephase.h]: and here

File: src/celengine/timelinephase.h [hunk 15]
```diff
@@ -78,48 +77,44 @@ class TimelinePhase
         return m_startTime <= t && t < m_endTime;
     }
 
-    static TimelinePhase* CreateTimelinePhase(Universe& universe,
+    static std::shared_ptr<const TimelinePhase> CreateTimelinePhase(Universe& universe,
```
Comment (375gnu) [inline src/celengine/timelinephase.h]: again

File: src/celengine/timelinephase.h [hunk 16]
```diff
@@ -78,48 +77,44 @@ class TimelinePhase
...
+    TimelinePhase& operator=(const TimelinePhase& phase) = delete;
 
 private:
     Body* m_body;
 
     double m_startTime;
     double m_endTime;
 
-    ReferenceFrame* m_orbitFrame;
+    std::shared_ptr<const ReferenceFrame> m_orbitFrame;
```
Comment (375gnu) [inline src/celengine/timelinephase.h]: again `const`

File: src/celengine/timelinephase.h [hunk 17]
```diff
@@ -78,48 +77,44 @@ class TimelinePhase
...
     Orbit* m_orbit;
-    ReferenceFrame* m_bodyFrame;
+    std::shared_ptr<const ReferenceFrame> m_bodyFrame;
```
Comment (375gnu) [inline src/celengine/timelinephase.h]: and again

File: src/celestia/celx_phase.cpp [hunk 18]
```diff
@@ -98,7 +73,7 @@ static int phase_timespan(lua_State* l)
 
     celx.checkArgs(1, 1, "No arguments allowed for to phase:timespan");
 
-    const TimelinePhase* phase = this_phase(l);
+    auto& phase = *this_phase(l);
```
Comment (375gnu) [inline src/celestia/celx_phase.cpp]: reference to pointer == pointer to pointer

File: src/celengine/body.cpp [hunk 19]
```diff
@@ -544,9 +544,9 @@ UniversalCoord Body::getPosition(double tdb) const
 {
     Vector3d position = Vector3d::Zero();
 
-    const TimelinePhase* phase = timeline->findPhase(tdb);
+    auto phase = timeline->findPhase(tdb);
```
Comment (375gnu) [inline src/celengine/body.cpp]: why did you remove `const` here

File: src/celengine/timelinephase.h (hunk 13 above)
Comment (pirogronian) [reply src/celengine/timelinephase.h]: I prefer to declare references as const whenever possible to avoid unneeded-modification-bugs.

File: src/celestia/celx_phase.cpp (hunk 18 above)
Comment (pirogronian) [reply src/celestia/celx_phase.cpp]: Actually this is a reference to shared_ptr.

File: src/celengine/body.cpp (hunk 19 above)
Comment (pirogronian) [reply src/celengine/body.cpp]: Probably because returned reference is const already.

Comment (pirogronian) [reply src/celengine/body.cpp]: Probably because returned reference is const already.

File: src/celengine/timelinephase.h (hunk 13 above)
Comment (375gnu) [reply src/celengine/timelinephase.h]: Celestia already has a bit more than allowed `const_cast`s :)

Comment (pirogronian) [reply src/celengine/timelinephase.h]: I didn't see any yet. I may start removing it in another PR.

Comment (375gnu) [reply src/celengine/timelinephase.h]: It's better not to introduce new

Comment (pirogronian) [reply src/celengine/timelinephase.h]: I dont, do I?

Comment (375gnu) [reply src/celengine/timelinephase.h]: Potentially with `const` put here and there once we will face a situation when api change or static_const is required.

Comment (pirogronian) [reply src/celengine/timelinephase.h]: Still feel const is better. If someone really need to modify particular data, then maybe it's a time to rethink general approach. Personally, I like to know exactly where and who modify particular data.

File: src/celengine/frametree.h [hunk 20]
```diff
@@ -92,15 +92,15 @@ class FrameTree
 private:
     Star* starParent;
     Body* bodyParent;
-    std::vector<TimelinePhase*> children;
+    std::vector<TimelinePhase::SharedConstPtr > children;
```
Comment (375gnu) [inline src/celengine/frametree.h]: extra space

File: src/celengine/observer.cpp [hunk 21]
```diff
@@ -57,7 +57,7 @@ static Vector3d slerp(double t, const Vector3d& v0, const Vector3d& v1)
 
 Observer::Observer()
 {
-    frame = new ObserverFrame();
+    frame = make_shared<ObserverFrame>();
```
Comment (375gnu) [inline src/celengine/observer.cpp]:
```
Observer::Observer() :
    frame(make_shared<ObserverFrame>())
{
...
```

File: src/celengine/observer.cpp [hunk 22]
```diff
@@ -83,13 +83,12 @@ Observer::Observer(const Observer& o) :
     locationFilter(o.locationFilter),
     displayedSurface(o.displayedSurface)
 {
-    frame = new ObserverFrame(*o.frame);
+    frame = make_shared<ObserverFrame>(*o.frame);
```
Comment (375gnu) [inline src/celengine/observer.cpp]: like above

File: src/celengine/observer.cpp [hunk 23]
```diff
@@ -816,25 +814,19 @@ void Observer::setFrame(ObserverFrame::CoordinateSystem cs, const Selection& ref
...
     {
-        auto* newFrame = new ObserverFrame(f);
-
-        if (frame != nullptr)
-        {
-            convertFrameCoordinates(newFrame);
-            delete frame;
-        }
-        frame = newFrame;
+        convertFrameCoordinates(f);
```
Comment (375gnu) [inline src/celengine/observer.cpp]: could you check `convertFrameCoordinates`? i'm not sure that your change is equivalent to the original code

File: src/celengine/observer.cpp [hunk 24]
```diff
@@ -1460,8 +1446,6 @@ ObserverFrame& ObserverFrame::operator=(const ObserverFrame& f)
 
 ObserverFrame::~ObserverFrame()
```
Comment (375gnu) [inline src/celengine/observer.cpp]: this can be moved to .h (= default)

File: src/celengine/observer.h [hunk 25]
```diff
@@ -66,30 +69,30 @@ class ObserverFrame
...
+    const ReferenceFrame::SharedConstPtr& getFrame() const;
 
     UniversalCoord convertFromUniversal(const UniversalCoord& uc, double tjd) const;
     UniversalCoord convertToUniversal(const UniversalCoord& uc, double tjd) const;
     Eigen::Quaterniond convertFromUniversal(const Eigen::Quaterniond& q, double tjd) const;
     Eigen::Quaterniond convertToUniversal(const Eigen::Quaterniond& q, double tjd) const;
 
-    static UniversalCoord convert(const ObserverFrame* fromFrame,
-                                  const ObserverFrame* toFrame,
+    static UniversalCoord convert(const ObserverFrame::SharedConstPtr &fromFrame,
```
Comment (375gnu) [inline src/celengine/observer.h]: please be consistent with `&` positions. in some places you have Foo &f, in other Foo& f

File: src/celengine/parseobject.cpp [hunk 26]
```diff
@@ -1307,7 +1307,7 @@ getFrameCenter(const Universe& universe, Hash* frameData, const Selection& defau
 }
 
 
-static BodyFixedFrame*
+static shared_ptr<const BodyFixedFrame>
```
Comment (375gnu) [inline src/celengine/parseobject.cpp]: BodyFixedFrame::Ptr

File: src/celestia/celx_phase.cpp (hunk 18 above)
Comment (375gnu) [inline src/celestia/celx_phase.cpp]: if it possible to put copy of shared_ptr to luauserdata? if not then it's better to put a raw pointer to data instead of the raw pointer to a smart pointer

Comment (pirogronian) [reply src/celestia/celx_phase.cpp]: That's how it's implemented already. Lua userdata is always presented to user as a raw pointer by default. So, by default we always have raw pointer to smart pointer. :-)

Comment (375gnu) [reply src/celestia/celx_phase.cpp]: okay

File: src/celengine/solarsys.cpp [hunk 27]
```diff
@@ -368,17 +363,15 @@ TimelinePhase* CreateTimelinePhase(Body* body,
         rotationModel = new ConstantOrientation(Quaterniond::Identity());
     }
 
-    TimelinePhase* phase = TimelinePhase::CreateTimelinePhase(universe,
+    auto phase = TimelinePhase::CreateTimelinePhase(universe,
                                                               body,
```
Comment (375gnu) [inline src/celengine/solarsys.cpp]: incorrect indentation

File: src/celengine/solarsys.cpp [hunk 28]
```diff
@@ -407,7 +400,7 @@ Timeline* CreateTimelineFromArray(Body* body,
         bool isFirstPhase = iter == timelineArray->begin();
         bool isLastPhase = *iter == timelineArray->back();
 
-        TimelinePhase* phase = CreateTimelinePhase(body, universe, phaseData,
+        auto phase = CreateTimelinePhase(body, universe, phaseData,
```
Comment (375gnu) [inline src/celengine/solarsys.cpp]: incorrect indentattion

File: src/celengine/solarsys.cpp [hunk 29]
```diff
@@ -629,22 +617,18 @@ static bool CreateTimeline(Body* body,
...
-            delete defaultOrbitFrame;
-            delete orbitFrame;
             delete rotationModel;
             return false;
         }
 
         // We finally have an orbit, rotation model, frames, and time range. Create
         // the object timeline.
-        TimelinePhase* phase = TimelinePhase::CreateTimelinePhase(universe,
+        auto phase = TimelinePhase::CreateTimelinePhase(universe,
```
Comment (375gnu) [inline src/celengine/solarsys.cpp]: incorrect indentation

File: src/celengine/timelinephase.cpp [hunk 30]
```diff
@@ -126,12 +80,12 @@ TimelinePhase::CreateTimelinePhase(Universe& universe,
         return nullptr;
     }
 
-    TimelinePhase* phase = new TimelinePhase(body,
+    auto phase = make_shared<const TimelinePhase>(body,
```
Comment (375gnu) [inline src/celengine/timelinephase.cpp]: indentation

File: src/celengine/timelinephase.h [hunk 31]
```diff
@@ -78,48 +78,41 @@ class TimelinePhase
         return m_startTime <= t && t < m_endTime;
     }
 
-    static TimelinePhase* CreateTimelinePhase(Universe& universe,
+    static TimelinePhase::SharedConstPtr CreateTimelinePhase(Universe& universe,
```
Comment (375gnu) [inline src/celengine/timelinephase.h]: indentation

File: src/celestia/celx_phase.cpp [hunk 32]
```diff
@@ -211,15 +186,14 @@ static int phase_gc(lua_State* l)
 {
     CelxLua celx(l);
 
-    PhaseReference* ref = static_cast<PhaseReference*>(celx.checkUserData(1, Celx_Phase));
-    if (ref == nullptr)
+    auto & ref = *this_phase(l);
+    if (ref)
```
Comment (375gnu) [inline src/celestia/celx_phase.cpp]: incorrect

should be smth like
```
auto t = this_phase(l);
if (t == nullptr)
{
   auto& ref = *t;
``

File: src/celestia/celx_phase.h [hunk 33]
```diff
@@ -12,10 +12,17 @@
...
 class TimelinePhase;
 
+inline int celxClassId(const std::shared_ptr<const TimelinePhase>&)
+{
+    return Celx_Phase;
+}
+
 extern void CreatePhaseMetaTable(lua_State* l);
-extern int phase_new(lua_State* l, const TimelinePhase& phase);
+extern int phase_new(lua_State* l,  const shared_ptr<const TimelinePhase>& phase);
```
Comment (375gnu) [inline src/celestia/celx_phase.h]: extra space before const

File: src/celestia/celx_phase.h [hunk 34]
```diff
@@ -12,10 +12,17 @@
 #ifndef _CELX_PHASE_H_
 #define _CELX_PHASE_H_
 
+#include <memory>
+
 struct lua_State;
 class TimelinePhase;
 
+inline int celxClassId(const std::shared_ptr<const TimelinePhase>&)
```
Comment (375gnu) [inline src/celestia/celx_phase.h]: i don't see where this fn is used

File: src/celestia/celx.cpp [hunk 35]
```diff
@@ -1851,7 +1851,7 @@ void CelxLua::newFrame(const ObserverFrame& f)
     frame_new(m_lua, f);
 }
 
-void CelxLua::newPhase(const TimelinePhase& phase)
+void CelxLua::newPhase(const shared_ptr<const TimelinePhase>& phase)
```
Comment (375gnu) [inline src/celestia/celx.cpp]: not changed to TimelinePhase::Ptr

File: src/celengine/solarsys.cpp (hunk 27 above)
Comment (pirogronian) [reply src/celengine/solarsys.cpp]: What do you mean?

File: src/celengine/timelinephase.h (hunk 31 above)
Comment (pirogronian) [reply src/celengine/timelinephase.h]: Again, what do you mean?

File: src/celestia/celx_phase.cpp (hunk 32 above)
Comment (pirogronian) [reply src/celestia/celx_phase.cpp]: `this_phase` already checks for nullptr.

File: src/celestia/celx_phase.h (hunk 34 above)
Comment (pirogronian) [reply src/celestia/celx_phase.h]: In `celx_internal.h` at line 316.

File: src/celengine/solarsys.cpp (hunk 27 above)
Comment (375gnu) [reply src/celengine/solarsys.cpp]: `body` should start just under `universe`

File: src/celengine/timelinephase.h (hunk 31 above)
Comment (375gnu) [reply src/celengine/timelinephase.h]: all arguments if placed on several lines should start on the same column

File: src/celestia/celx_phase.cpp (hunk 32 above)
Comment (375gnu) [reply src/celestia/celx_phase.cpp]: in this case this code is too obscure. references and smart pointers pointing to nullptr is a bad practice

Comment (pirogronian) [reply src/celestia/celx_phase.cpp]: Even if program should never enter this code with null pointer? Maybe `assert(t != nullptr)`?

Comment (pirogronian) [reply src/celestia/celx_phase.cpp]: @375gnu So, are you sure I should do additional check in every place i get pointer to data? And should it be `if { }` statement or just `assert` would be enough?

Comment (375gnu) [reply src/celestia/celx_phase.cpp]: I'd prefer `assert`

**Comment by pirogronian** [comment]
I can't see any remained notice...

**Comment by 375gnu** [comment]
@pirogronian 
https://github.com/CelestiaProject/Celestia/pull/344/files/3a05fa914f88ecabe292ccf3d884da196418c719#diff-7be42549281eeed64f5a9370badd11c8L521

**Comment by 375gnu** [comment]
@pirogronian now it looks good, could you cleanup commit history?

**Comment by 375gnu** [comment]
@pirogronian why did you merge dirty history?

**Comment by pirogronian** [comment]
I was too fast, sorry. Could you revert it? I will later restore this branch and clean it.

**Comment by 375gnu** [comment]
@pirogronian I already merged all your commits into a single one
